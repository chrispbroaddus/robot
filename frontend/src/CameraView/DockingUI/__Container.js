import React from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components'
import { bindActionCreators } from 'redux'
import { push } from 'react-router-redux'
import { connect } from 'react-redux'
import './docking-ui.styles.scss'
import * as vehicleDucks from '../../redux/modules/vehicles'
import DockingAnimation from './DockingAnimation'
import TelemetryData from './DockingAnimation/TelemetryData.js'

const Container = styled.div`
  position: absolute;
  right: 55px;
  top: 260px;
  z-index: 500;
`

class Docking extends React.Component {
  dock

  static propTypes = {
    message: PropTypes.object,
    angle: PropTypes.number,
    distanceX: PropTypes.number,
    distanceY: PropTypes.number,
    status: PropTypes.string,
    setDockingDivLocation: PropTypes.func.isRequired
  }

  constructor(props) {
    super(props)
    this.state = {
      locationOfVehicle: {
        X: 0.0,
        Y: 0.0
      }
    }
  }

  componentWillUpdate = prevProps => {
    // console.log('did update', this.props.distanceX, this.props.distanceY)
    if (prevProps === this.props) {
      return
    }

    let location = this.computeLocationOfVehicle()
    this.setState({
      locationOfVehicle: location
    })
  }

  computeLocationOfVehicle = () => {
    const { distanceX, distanceY, angle } = this.props
    /* ------------------- Compute the Center of the vehicle-----------
    Xb = Xv - w/2*sin(angle)
    Yb = Yv + w/2*cos(angle)

    where Xv and Yv is distanceX,distanceY from telemetry data and w is the approximate meter-dimentions of the robot's width and height. */
    const robotDimentionsMeters = 0.2612
    let Xb = distanceX - robotDimentionsMeters * Math.sin(angle)
    let Yb = distanceY + robotDimentionsMeters * Math.cos(angle)
    /* ------------------- Compute the Location of the vehicle-----------
    X = Px/A * (Xb -B) + D
    Y = Px/A * (Yb -C) + E

    - where Px is dimentions in pixels of animtion container, (268px * 268px for now)
    - where A is dimentions in meters of robot container space (for now its 2m * 2m)
    - (Xb,Yb) is the center coordinate of the vehicle
    - (D,E) is the top left location of the div within the browser window
    */
    // + (dockingDivLocation[0]
    // + (dockingDivLocation[1]
    const pixelConversionOfContainer = 134
    return {
      X: pixelConversionOfContainer * (Xb - -1) - 35,
      Y: pixelConversionOfContainer * (Yb - -0.2) - 35
    }
  }

  render() {
    const {
      angle,
      distanceX,
      distanceY,
      status,
      setDockingDivLocation
    } = this.props
    const { locationOfVehicle } = this.state

    return (
      <Container>
        <TelemetryData
          distanceX={distanceX}
          distanceY={distanceY}
          angle={angle}
        />
        <DockingAnimation
          angle={angle}
          distanceX={distanceX}
          distanceY={distanceY}
          status={status}
          locationOfVehicle={locationOfVehicle}
          setDockingDivLocation={setDockingDivLocation}
        />
      </Container>
    )
  }
}

// map redux store to component props
const mapStateToProps = state => {
  return {
    dockingDivLocation: state.dockingDivLocation
  }
}

const mapDispatchToProps = dispatch => {
  return bindActionCreators({ ...vehicleDucks, push }, dispatch)
}

export default connect(mapStateToProps, mapDispatchToProps)(Docking)
