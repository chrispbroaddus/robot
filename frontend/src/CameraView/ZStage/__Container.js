import React from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components'
import Modal from 'react-modal'
import Close from 'react-icons/lib/md/close'
import ZStageBody from './Body'
import ZStageLeg from './Leg'
import { scaleAtoB } from '../../utils/translations'

const Container = styled.div`
  position: absolute;
  bottom: 25px;
  left: 40px;
  z-index: 1000;
  padding: 15px;
  border-radius: 20px;
  background-color: rgba(0, 0, 0, 0.5);
`
const ModalContainer = styled.div`
  position: relative;
  margin: auto;
  user-select: none;
`
const CloseButton = styled(Close)`
  color: #fe7867;
  font-size: 15px;
  position: absolute;
  top: -5px;
  right: -5px;
  cursor: pointer;
`
const Title = styled.h5`
  text-transform: uppercase;
  letter-spacing: 2px;
  color: white;
  margin: 0 0 15px 0;
  padding: 0 10px;
`
const Zippy = styled.div`
  height: 355px;
  padding: 0 35px;
  display: flex;
  flex-direction: row;
  justify-content: center;
  align-items: flex-end;
`
const Leg = styled(ZStageLeg)`display: inline-block;`
const BodyContainer = styled.div`
  height: 300px;
  & > svg {
    cursor: grab;
    // transition: 0.1s;
    -webkit-backface-visibility: hidden;
    position: relative;
    top: 0;
    left: 0;
  }
`

const HeightIndicator = styled.div`
  color: white;
  text-transform: uppercase;
  text-align: center;
  position: relative;
  top -15px;
  font-size: 11px;

  & > span {
    color: #00E4B9;
  }
`

const Instructions = styled.div`
  font-size: 12px;
  color: white;
  text-align: center;
  margin: 15px 0 7px;
`

// Adjust the below to change the effective range of the mousemove
const containerMin = 30
const containerMax = 200
// Adjust the below to change the maximum up/down offsets for the robot
// SVG
const pictureMin = -75
const pictureMax = 40
// Adjust the below to change the minimum/maximum values that the vehicle
// can accept
const vehicleMin = 0
const vehicleMax = 0.4
// Adjust the below to change the minimum/maximum height from the ground
const groundHeightMin = 5.5
const groundHeightMax = 21.5

class ZStageContainer extends React.PureComponent {
  static propTypes = {
    onSendValue: PropTypes.func.isRequired,
    isOpen: PropTypes.bool,
    onRequestClose: PropTypes.func
  }

  static defaultProps = {
    isOpen: false,
    onRequestClose: () => {}
  }

  constructor(props) {
    super(props)

    this.mouseDown = false
    this.state = {
      zStageValue: vehicleMin.toFixed(2),
      groundHeight: groundHeightMin.toFixed(1)
    }
  }

  handleMouseDown = evt => {
    this.mouseDown = true
    this.bodySvgRef.style.cursor = '-webkit-grabbing'
  }

  handleMouseUp = () => {
    const { onSendValue } = this.props
    const { zStageValue } = this.state

    this.mouseDown = false
    this.bodySvgRef.style.cursor = '-webkit-grab'
    onSendValue(zStageValue)
  }

  handleMouseMove = evt => {
    if (this.mouseDown === true) {
      this.bodySvgRef.style.cursor = '-webkit-grabbing'
      const container = this.bodyContainerRef.getBoundingClientRect()
      // const body = this.bodySvgRef.getBoundingClientRect()
      // console.log(`Cont:   left\t${container.left}, top\t${container.top} `)
      // console.log(`Body:   left\t${body.left}, top\t${body.top} `)
      // console.log(`Mouse:  x\t${evt.clientX}, y\t\t${evt.clientY}`)
      const delta = evt.clientY - container.top
      // console.log(`ydelta ${delta}`)

      if (delta > containerMin && delta < containerMax) {
        const bodyTop = scaleAtoB(
          delta,
          containerMin,
          containerMax,
          pictureMin,
          pictureMax
        )
        this.bodySvgRef.style.top = `${Math.ceil(bodyTop)}px`

        const zStage = scaleAtoB(
          Math.ceil(bodyTop),
          pictureMin,
          pictureMax,
          vehicleMax,
          vehicleMin
        )

        const groundHeight = scaleAtoB(
          zStage,
          vehicleMin,
          vehicleMax,
          groundHeightMin,
          groundHeightMax
        )

        this.setState({
          zStageValue: zStage.toFixed(2),
          groundHeight: groundHeight.toFixed(1)
        })
      } else if (delta < containerMin) {
        // If the mouse is at the extremes, assume the max/min values
        this.setState({
          zStageValue: vehicleMax.toFixed(2),
          groundHeight: groundHeightMax.toFixed(1)
        })
        this.bodySvgRef.style.top = `${pictureMin}px`
      } else if (delta > containerMin) {
        this.setState({
          zStageValue: vehicleMin.toFixed(2),
          groundHeight: groundHeightMin.toFixed(1)
        })
        this.bodySvgRef.style.top = `${pictureMax}px`
      }
    }
  }

  render() {
    const { isOpen, onRequestClose } = this.props
    const { groundHeight } = this.state

    return (
      <Container>
        <Modal
          isOpen={isOpen}
          onRequestClose={onRequestClose}
          className="zstage-modal"
          overlayClassName="modal-overlay"
          contentLabel="Z Stage"
        >
          <ModalContainer>
            <CloseButton onClick={onRequestClose} />
            <Title>Z-Stage</Title>
            <Zippy>
              <Leg />
              <BodyContainer
                onMouseDown={this.handleMouseDown}
                onMouseUp={this.handleMouseUp}
                innerRef={r => (this.bodyContainerRef = r)}
              >
                <ZStageBody
                  onMouseMove={this.handleMouseMove}
                  innerRef={r => (this.bodySvgRef = r)}
                />
              </BodyContainer>
              <Leg />
            </Zippy>
            <HeightIndicator>
              <span>{groundHeight}</span> in
            </HeightIndicator>
            <Instructions>Drag center of robot to position</Instructions>
          </ModalContainer>
        </Modal>
      </Container>
    )
  }
}

export default ZStageContainer
