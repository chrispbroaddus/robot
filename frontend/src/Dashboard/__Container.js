import React from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components'
import { connect } from 'react-redux'
import { bindActionCreators } from 'redux'
import { push } from 'react-router-redux'
import * as vehicleDucks from '../redux/modules/vehicles'
import * as notificationDucks from '../redux/modules/notifications'
import CommonShapes from '../Shared/CommonShapes'
import Navigation from './Navigation'
import Title from './Title'
import VehicleCard from './VehicleCard'
import AddSimulatorCard from './AddSimulatorCard'
import WebSocketClient from '../utils/WebSocket'

const Container = styled.div`
  background: #f5f7fa;
  min-height: 100vh;
`

const VehicleContainer = styled.div`
  display: flex;
  flex-flow: row wrap;
  margin-left: 82px;
  margin-top: 20px;
  transition: 0.2s all;
`

export class Dashboard extends React.Component {
  static propTypes = {
    addVehicle: PropTypes.func.isRequired,
    editVehicle: PropTypes.func.isRequired,
    fetchVehicles: PropTypes.func.isRequired,
    fetchAndUpdateVehicle: PropTypes.func.isRequired,
    killSimulator: PropTypes.func.isRequired,
    setVehicleStatus: PropTypes.func.isRequired,
    getLatestCameraFrame: PropTypes.func.isRequired,
    push: PropTypes.func.isRequired,
    requestControl: PropTypes.func.isRequired,
    addNotification: PropTypes.func.isRequired,
    vehicles: PropTypes.arrayOf(CommonShapes.Vehicle).isRequired
  }

  constructor(props) {
    super(props)

    this.ws = null
  }

  componentDidMount() {
    this.props.fetchVehicles()
  }

  componentWillMount() {
    if (this.ws) this.ws.closeSocket()
  }

  /**
   * Creates a placeholder vehicle until we get a "simulator up" message in the
   * notify socket
   */
  addPlaceholderVehicle = (
    id,
    status = vehicleDucks.VEHICLE_STATUS.LAUNCHING,
    cameras = []
  ) => {
    const { VEHICLE_TYPES } = vehicleDucks
    const vehicle = {
      id,
      status,
      vehicle_id: id,
      cameras,
      location: {
        location: {
          lat: 37.3382,
          lon: 121.8863,
          alt: 12345
        },
        timestamp: new Date().toUTCString()
      },
      operator: {
        name: ''
      },
      type: VEHICLE_TYPES.SIMULATOR
    }

    this.props.addVehicle(vehicle)
    this.openNotifyWS(id)
  }

  openNotifyWS = id => {
    this.ws = new WebSocketClient(
      this,
      `ws${window.zippyconfig.scheme}://${window.zippyconfig
        .backend}/api/v1/ws/simulator/${id}/notifier`,
      {
        name: 'SimulatorNotifier',
        onmessage: this.handleMessage
      }
    )
  }

  handleMessage = msg => {
    const message = JSON.parse(msg.data)

    // Other statuses could be 'start' and 'ping', neither of which we care
    // about at the moment
    if (message.status === 'ready') {
      this.props.fetchAndUpdateVehicle(message.id)
      this.ws.closeSocket()
    } else if (message.status === 'error') {
      this.props.setVehicleStatus(message.id, 'disabled')
      this.ws.closeSocket()
    }
  }

  handleClickOperate = (vehicleId, camera) => {
    const { push } = this.props

    push(`/app/vehicle/${vehicleId}/${camera}`)
  }

  handleClickShutdownSimulator = vehicleId => {
    const { killSimulator, push } = this.props
    killSimulator(vehicleId)
      .then(() => {
        push('/')
      })
      .catch(error => {
        // this.setState({ disabled: false })
      })
  }

  render() {
    const { vehicles, getLatestCameraFrame } = this.props
    const { getFrontCamera, VEHICLE_TYPES } = vehicleDucks

    return (
      <Container>
        <Navigation />
        <Title />
        <VehicleContainer>
          {vehicles.map(vehicle => {
            const frontCamera = getFrontCamera(vehicle)
            const role = frontCamera ? frontCamera.role : null

            return (
              <VehicleCard
                key={vehicle.id}
                vehicleId={vehicle.id}
                // Why is this called realVehicleId? Because currently we are
                // using sessionId in place of vehicleId everywhere. This works
                // well for the most part, but it's a stopgap until point and go
                // can actually understand vehicle_ids.
                realVehicleId={vehicle.vehicle_id}
                cameraRole={role}
                status={vehicle.status}
                type={vehicle.type ? vehicle.type : VEHICLE_TYPES.VEHICLE}
                getLatestCameraFrame={getLatestCameraFrame}
                handleClickOperate={this.handleClickOperate}
                handleClickShutdown={this.handleClickShutdownSimulator}
              />
            )
          })}
          <AddSimulatorCard onComplete={this.addPlaceholderVehicle} />
        </VehicleContainer>
      </Container>
    )
  }
}

const mapStateToProps = state => {
  return {
    vehicles: state.vehicles.vehicles
  }
}

const mapDispatchToProps = dispatch => {
  return bindActionCreators(
    { ...vehicleDucks, ...notificationDucks, push },
    dispatch
  )
}

export default connect(mapStateToProps, mapDispatchToProps)(Dashboard)
