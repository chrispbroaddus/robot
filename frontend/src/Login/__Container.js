import React from 'react'
import styled from 'styled-components'
import PropTypes from 'prop-types'
import { push } from 'react-router-redux'
import { bindActionCreators } from 'redux'
import { connect } from 'react-redux'
import LoginDetailsContainer from './LoginDetails'
import Logo from '../Shared/Logo'
import * as userDucks from '../redux/modules/user'
import ZippyIllustration from '../images/loginZippyIllustration.png'

const Container = styled.div`
  position: absolute;
  height: 100vh;
  width: 100vw;
  background-color: white;
  background-image: url('${ZippyIllustration}');
  background-size: cover;
`

const LogoContainer = styled(Logo)`
  margin-left: 43px;
  margin-top: 28px;
  display: block;
`

export const Login = props =>
  <Container>
    <LogoContainer />
    <LoginDetailsContainer {...props} />
  </Container>

Login.propTypes = {
  login: PropTypes.func.isRequired,
  push: PropTypes.func.isRequired,
  isAuthenticated: PropTypes.bool.isRequired
}

Login.defaultProps = {}

const mapStateToProps = state => {
  return {
    isAuthenticated: state.user.isAuthenticated
  }
}

const mapDispatchToProps = dispatch => {
  return bindActionCreators({ ...userDucks, push }, dispatch)
}

export default connect(mapStateToProps, mapDispatchToProps)(Login)
