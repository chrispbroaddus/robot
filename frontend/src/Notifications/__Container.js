import React from 'react'
import PropTypes from 'prop-types'
import styled, { keyframes } from 'styled-components'
import { bindActionCreators } from 'redux'
import { push } from 'react-router-redux'
import { connect } from 'react-redux'
import WebSocketClient from '../utils/WebSocket'
import * as notificationDucks from '../redux/modules/notifications'
import CommonShapes from '../Shared/CommonShapes'

const heightPx = 60

const AnimateIn = keyframes`
  from {
    height: 0;
    opacity: 0;
  }

  to {
    height: ${heightPx}px;
    opacity: 1;
  }
`

const AnimateOut = keyframes`
  from {
    height: ${heightPx}px;
    opacity: 1;
  }

  100% {
    height: 0;
    opacity: 0;
  }
`

const Container = styled.div`
  height: 0; // Initially hide the banner
  opacity: 0;
  background-color: ${props => (props.color ? props.color : '#2b2b2b')};
  width: 100%;
  z-index: 1;
  overflow: hidden;
  animation: 0.2s ease-in forwards
    ${props => (props.visible ? AnimateIn : AnimateOut)};
  cursor: pointer;
`

const InnerContainer = styled.div`
  padding-top: 22px;
  color: ${props => (props.color ? props.color : '#fff')};
  font-family: 'Montserrat-Regular';
  letter-spacing: 1px;
  font-size: 13px;
  text-align: center;
`

const NOTIFY_WS_URL = `ws${window.zippyconfig.scheme}://${window.zippyconfig
  .backend}/api/v1/ws/operator/notify`

export class NotificationContainer extends React.Component {
  static propTypes = {
    visible: PropTypes.bool,
    notifications: PropTypes.arrayOf(PropTypes.object).isRequired,
    addNotification: PropTypes.func.isRequired,
    dismissNotification: PropTypes.func.isRequired,
    currentNotification: CommonShapes.Notifications
  }

  static defaultProps = {
    visible: false
  }

  constructor(props) {
    super(props)
    this.timeout = null
    this.notificationWS = null
  }

  componentDidMount() {
    this.notificationWS = new WebSocketClient(this, NOTIFY_WS_URL, {
      name: 'Notifications',
      onmessage: this.handleWSMessage
    })
  }

  componentWillUpdate(nextProps) {
    const { dismissNotification, currentNotification } = nextProps
    if (currentNotification !== null) {
      if (currentNotification.timeout > 0) {
        if (this.timeout) clearTimeout(this.timeout)
        this.timeout = setTimeout(
          dismissNotification,
          currentNotification.timeout
        )
      }
    }
  }

  componentWillUnmount() {
    if (this.notificationWS) this.notificationWS.closeSocket()
  }

  /**
   * Handles the messages coming from the notification websocket
   * @param  {Object} message Message from the ws
   */
  handleWSMessage = msg => {
    const { addNotification } = this.props
    const message = JSON.parse(msg.data)
    // Expecting the message to take the shape:
    //   message: {
    //     type: String,
    //     message: String,
    //     payload: String,
    //   }

    if (message.type === 'vehicle_disconnected') {
      // Display a message indefinitely when disconnected
      addNotification(
        message.message,
        notificationDucks.NOTIFICATION_STATUS.DANGER,
        0
      )
    }
  }

  handleClick = () => {
    if (this.timeout) clearTimeout(this.timeout)
    this.props.dismissNotification()
  }

  render() {
    const { visible, currentNotification } = this.props
    let bgColor = null,
      fgColor = null
    if (currentNotification) {
      const { bgColor: bg, fgColor: fg } = notificationDucks.notificationColor(
        currentNotification.status
      )
      bgColor = bg
      fgColor = fg
    }

    return (
      <Container
        visible={visible}
        onClick={this.handleClick}
        color={bgColor}
        title="Click to dismiss notification"
      >
        <InnerContainer color={fgColor}>
          {currentNotification && currentNotification.message}
        </InnerContainer>
      </Container>
    )
  }
}

const mapStateToProps = state => {
  const notifications = state.notifications.notifications
  let currentNotification
  if (notifications.length > 0) {
    currentNotification = notifications[0]
  } else {
    currentNotification = null
  }
  return {
    notifications: notifications,
    currentNotification,
    visible: notifications.length > 0
  }
}

const mapDispatchToProps = dispatch => {
  return bindActionCreators({ ...notificationDucks, push }, dispatch)
}

export default connect(mapStateToProps, mapDispatchToProps)(
  NotificationContainer
)
