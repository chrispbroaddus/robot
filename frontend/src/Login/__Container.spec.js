import React from 'react'
import { shallow } from 'enzyme'
import toJson from 'enzyme-to-json'
import { Login } from './__Container'

describe('Login', () => {
  let props

  beforeEach(() => {
    props = {
      login: () => {},
      push: () => {},
      isAuthenticated: false,
      location: {
        state: {}
      }
    }
  })

  it('should render self and subcomponents', () => {
    const wrapper = shallow(<Login {...props} />)
    expect(toJson(wrapper)).toMatchSnapshot()
  })
})
