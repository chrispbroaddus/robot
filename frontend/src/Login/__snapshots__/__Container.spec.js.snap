// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login should render self and subcomponents 1`] = `
<styled.div>
  <Styled(Logo) />
  <LoginDetails
    isAuthenticated={false}
    location={
      Object {
        "state": Object {},
      }
    }
    login={[Function]}
    push={[Function]}
  />
</styled.div>
`;
